For #467: Created astromancer lunar embrace card

paulodamaso opened the pull request:
Closes #467: Created astromancer lunar embrace card

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/Heal.cs:
```diff
@@ -0,0 +1,33 @@
+        }
+        else
+        {
+            Debug.Log("Healing...");
```
Needed?

paulodamaso replied:
@SilasReinagel Sorry

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/Heal.cs:
```diff
@@ -0,0 +1,33 @@
+        else
+        {
+            Debug.Log("Healing...");
+            target.Members[0].State.GainHp(_amount);
```
No indexing.

paulodamaso replied:
@SilasReinagel Done

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/EffectType.cs:
```diff
@@ -20,4 +20,7 @@ public enum EffectType
     EvadeAttacks = 19,
+    Regenerate = 20,
+    BuffStrengthFlat = 21,
+    BuffToughnessFlat = 21
```
You can't have two `21`s

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/AllEffects.cs:
```diff
@@ -30,6 +30,10 @@ public static class AllEffects
         { EffectType.InterceptAttackForTurns, e => new ForNumberOfTurns(new InterceptAttack(), e.NumberOfTurns)},
         { EffectType.Attack, e => new Attack(e.FloatAmount)},
         { EffectType.EvadeAttacks, e => new Recurrent(new Evade(), e.IntAmount) },
+        { EffectType.Regenerate, e => new ForNumberOfTurns(new Heal(e.IntAmount), e.NumberOfTurns) },
```
Does this trigger the effect at the start of each turn?

Should have a test to ensure that the character is correctly healed as turns advance.

SilasReinagel replied:
`ForNumberOfTurns` only keeps passive effects.

It doesn't apply Active effects.

No-one seems to be triggering this. 

This composition looks incomplete. If it is complete, then it can be proved with a test. But it looks dead wrong and and non-functional.

SilasReinagel commented:
PR should never be submitted if program doesn't compile. This wastes time needlessly.

![image](https://user-images.githubusercontent.com/10834339/68432996-c28ad780-0172-11ea-9bdf-655120429fac.png)

paulodamaso commented:
@SilasReinagel Sorry, I mistakenly pushed some development code. Fixed now.

paulodamaso commented:
@SilasReinagel Fixed some things:
- `Heal` now works, it is triggered on start of the turn
- Improved the turn passing mechanism, making it with events and uncoupling it from memberstate. 
- Added `OnTurnStart`and `OnTurnEnd ` events
- Fixed tests for `ForNumberOfTurns` and added tests for `HealFlatForTurnsOnTurnStart`

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/EffectOnTurnStart.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        _performer = source;
+        _effectTarget = target;
+        BattleEvent.Subscribe<TurnStart>((turnStart) => Execute(), this);
```
This must get cleaned up to prevent Memory Leak.

Need unsubscribe logic.

paulodamaso replied:
@SilasReinagel Please see the other one about unsubscriving events

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/EffectOnTurnStart.cs:
```diff
@@ -0,0 +1,30 @@
+ */
+public class EffectOnTurnStart : Effect
+{
+    private Effect _effect;
```
No reason to store fields.

paulodamaso replied:
@SilasReinagel remove them

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/EffectOnTurnStart.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        _performer = source;
+        _effectTarget = target;
+        BattleEvent.Subscribe<TurnStart>((turnStart) => Execute(), this);
```
For a lambda method, if the input parameter is unused, convention says you should use and underscore.

`_ => Execute()`

paulodamaso replied:
@SilasReinagel Done

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/EffectOnTurnStart.cs:
```diff
@@ -0,0 +1,30 @@
+
+    public void Apply(Member source, Target target)
+    {
+        _performer = source;
```
These should be passed in the lambda, rather than stored in the class.

`BattleEvent.Subscribe<TurnStart>(_ => Execute(source, target), this);`

paulodamaso replied:
@SilasReinagel Done

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/EffectOnTurnStart.cs:
```diff
@@ -0,0 +1,30 @@
+
+    void Execute()
+    {
+        Debug.Log("Turn started!");
```
Remove the logging.

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/ForNumberOfTurns.cs:
```diff
@@ -1,38 +1,32 @@
+    public bool IsActive => _remainingDuration >= 0;
 
-    public IStats Stats { get; } = new StatAddends();
-    public bool IsDebuff => _isDebuff;
```
How will poison get cleansed if Temporal State is no longer implemented? 

This class still needs to implement Temporal State.

paulodamaso replied:
@SilasReinagel Poison isn't implemented yet; I though in implementing it as an effect that causes damage over time.

SilasReinagel replied:
We can't take away the IsDebuff flag, since that breaks the existing effect `CleanseDebuffs`

SilasReinagel replied:
I still don't understand why we stopped implementing `ITemporalState`. There seems to be no reason for it.

I will merge this and open a bug ticket.

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/ForNumberOfTurns.cs:
```diff
@@ -1,38 +1,32 @@
         _effect = effect;
         _remainingDuration = duration;
-        _isDebuff = isDebuff;
+        BattleEvent.Subscribe<TurnEnd>((turnEnd) => AdvanceTurn(), this);
```
All Event Subscriptions must have Unsubscriptions.

paulodamaso replied:
@SilasReinagel `Unsubscriptions` weren't needed until the last PR. We don't have them for any other classes that uses messaging subscription (`Attack`, `EffectOnAttacked`, EffectOnAttacker` etc,). Where should I put it?

SilasReinagel replied:
Well, for things that naturally expires, they should unsubscribe themselves once finished. Then the garbage collector can clean them up. 

Mentally, you should see a red flag any time there is no unsubscribe for any subscription.

I just ran into some program crashing bugs in another project due to missing unsubscriptions, so now I am being more vigilant about them.

paulodamaso replied:
@SilasReinagel Okay, fil add them to the related events

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/HealFlatForTurnsOnTurnStart.cs:
```diff
@@ -0,0 +1,16 @@
+        _amount = amount;
+    }
+
+    public void Apply(Member source, Target target)
```
For `Regen` itself, the caster who applied the effect is not the one who is doing the healing. The character with Regen heals himself.

paulodamaso replied:
@SilasReinagel `Heal` effect doesn't use the caster info. The target always heals itself:
```
    {
        target.Members.ForEach(
            member => {
                member.State.GainHp(_amount);
            }
        );
```

paulodamaso commented:
@SilasReinagel Fixed, please take a look
